### ISSUE TITLE
CI Caching: Cannot use store = /path/to/store in nix.conf

### ISSUE THREAD

**Issue by DrRuhe** [ipetkov/crane#154]
I'm trying to get gitlab caching to work with nix. 

Since their caching does not allow caching global paths (so no caching /nix/store directly) I want to use a "local" nix store that can be cached by their CI system. For that I took inspiration from [this project to cache nix store in gitlab ci](https://gitlab.com/srghma/nix-docker-cached-store-reproduction) where they use the local folder `./.mycache` as a nix store. 

However when trying to build a crane flake using this store, errors pop up for `cargoHelperFunctionsHook`

- this is inside the nixos/nix docker container
- I mounted my local project in the workdir, which is just the default crane template
- I setup the `.mycache` directory like [this](https://gitlab.com/srghma/nix-docker-cached-store-reproduction/-/blob/master/setup-ci.sh)
```
bash-5.1# nix --extra-experimental-features nix-command --extra-experimental-features flakes build -L .
warning: Git tree '/project' is dirty
bash-5.1# echo "store = $(pwd)/.mycache" > /etc/nix/nix.conf
bash-5.1# nix --extra-experimental-features nix-command --extra-experimental-features flakes build -L .
warning: Git tree '/project' is dirty
cargoHelperFunctionsHook> cp: cannot create regular file '/nix/store/b9dfvim2aa1yfndx8zyqrdpgp3m1p8cq-cargoHelperFunctionsHook/nix-support/setup-hook': Permission denied
error: builder for '/nix/store/q76lxyi5cp00l310jyj1jrx77cp98p9d-cargoHelperFunctionsHook.drv' failed with exit code 1
error: 1 dependencies of derivation '/nix/store/8nwh5zpw6rvhi190rrm9znfxlwkw675a-quick-start-0.1.0.drv' failed to build
bash-5.1# 
```

It seems like some script is trying to copy stuff inside the store without honoring the `store` option set in the config.

**Comment by ipetkov** [comment]
@DrRuhe how are you installing Nix on the runner? Have you tried setting `nix.conf` _before_ the installation?

From what I've seen, a multi-user install requires restarting the daemon process to pick up changes to `nix.conf`. Judging from the error it looks like the updated store path isn't picked up everywhere. Don't think this is specific to `cargoHelperFunctionsHook` as it doesn't even have any dependencies but looks like it is failing when trying to store its own output...

It is worth mentioning that successfully configuring a different store path will require rebuilding _the entire universe_, meaning you won't even get to benefit from the nixos.org caches. If that isn't an acceptable direction, I'd recommend trying to set up something like [cachix](https://cachix.org)

**Comment by DrRuhe** [comment]
TL:DR: For now I'll use cachix, but I'd like an alternative that's not 39€/month for a private store. 

So, the reason I got to this point was the following:

- I want to use nix as the build tool for rust projects
- I don't like setting up CI, so I'd like to have a config that just uses the stuff already present inside a nix flake
- I want to use the checks in the nix flake in the CI steps of GitLab
- CI steps should also cache the nix store since otherwise, it rebuilds on every push

To do that I created the following `.gitlab-ci.yml`:
```
stages:
  - build
  - check

cache:
  key: nix-store
  paths:
    - /nix/

.nix-build:
  stage: check
  image: nixos/nix
  script:
    - nix --extra-experimental-features nix-command --extra-experimental-features flakes build -L $CHECK
  variables:
    CHECK: "."


build:
  extends: .nix-build
  stage: build


cargo-fmt:
  extends: .nix-build
  variables:
    CHECK: ".#checks.x86_64-linux.my-crate-fmt"

cargo-clippy:
  extends: .nix-build
  variables:
    CHECK: ".#checks.x86_64-linux.my-crate-clippy"


cargo-doc:
  extends: .nix-build
  variables:
    CHECK: ".#checks.x86_64-linux.my-crate-doc"

cargo-nextest:
  extends: .nix-build
  variables:
    CHECK: ".#checks.x86_64-linux.my-crate-nextest"

cargo-tarpaulin:
  extends: .nix-build
  variables:
    CHECK: ".#checks.x86_64-linux.my-crate-coverage"
```

This neatly works to use the preexisting nix flake to check everything, and even gives different jobs for each check so the CI job that failed directly tells me which check did not succeed. However, everything has to be reinstalled every time.

For each of the jobs, it will rebuild everything since the nix store does not get persisted to cache intermediate results. 

From your comments, I conclude that caching the nix store by placing it in a cacheable location is not the way. For now, the only alternative I see is using cachix like you mentioned, however, I am not a big fan of their pricing models. 

@ipetkov Do you know of alternatives that can be used to cache private stores for a less hefty price?

**Comment by ipetkov** [comment]
@DrRuhe another option could be exporting your store directly into an archive format, caching that, and then unpacking it before running the pipeline. There appears to be a [`nix-store --export`](https://nixos.org/manual/nix/stable/command-ref/nix-store.html#operation---export) operation, but I'm not sure how much or little effort it would take to use it in a script

**Comment by DrRuhe** [comment]
So after some more tinkering, I finally settled on just running all checks in a single stage. While this is not as ergonomic as possible it is good enough and indicates that something is off, which will be verified locally.  

> There appears to be a [`nix-store --export`](https://nixos.org/manual/nix/stable/command-ref/nix-store.html#operation---export) operation

I have tried this for a bit. Effectively I exported the relevant store paths into a file, used the gitlab CI caching for that, and imported that on each stage. But since on every run every relevant store path had to be first imported and later exported, it added minutes to each step in the CI (up to the point that simple checks would take ~5-6 minutes). While there is probably room for improvement by not exporting the entire store and only relative paths, this will probably take 2-3 minutes anyway.

For more advanced use cases cachix is probably the best option, just because it is relatively easy to set up.

Another (rather time-consuming) option would be setting up a custom gitlab-runner, but just for personal projects, it feels like too much of a hassle.

**Comment by ipetkov** [comment]
> While there is probably room for improvement by not exporting the entire store and only relative paths, this will probably take 2-3 minutes anyway.

@DrRuhe You could try diffing the store before and after the checks are run and only exporting the changes (IIRC that's what cachix's github action does by default) so you don't end up exporting whatever else is in the NixOS caches
